ktor-client/ktor-client-winhttp/windows/src/io/ktor/client/engine/winhttp/WinHttpClientEngine.kt

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.engine.winhttp
+
+import io.ktor.client.engine.*
+import io.ktor.client.engine.winhttp.internal.*
+import io.ktor.client.plugins.*
+import io.ktor.client.plugins.websocket.*
+import io.ktor.client.request.*
+import io.ktor.util.*
+import kotlinx.coroutines.*
+
+internal class WinHttpClientEngine(
+    override val config: WinHttpClientEngineConfig
+) : HttpClientEngineBase("ktor-winhttp") {
+
+    override val dispatcher: CoroutineDispatcher = Dispatchers.Unconfined
+
+    override val supportedCapabilities = setOf(HttpTimeout, WebSocketCapability)
+
+    private val session = WinHttpSession(config)
+
+    init {
+        coroutineContext.job.invokeOnCompletion {
+            session.close()
+        }
+    }
+
+    @OptIn(InternalAPI::class)
+    override suspend fun execute(data: HttpRequestData): HttpResponseData {
+        val callContext = callContext()
+        return try {
+            session.executeRequest(data, callContext)
+        } catch (cause: Throwable) {
+            callContext.cancel(CancellationException("Failed to execute request", cause))
+            throw cause
+        }
+    }
+}
+
+@Suppress("KDocMissingDocumentation")
+public class WinHttpIllegalStateException(cause: String) : IllegalStateException(cause)
```

Comment:
why it's needed?

Reply:
It's used in the same cases as a similar `CurlIllegalStateException` in curl engine, e.g. https://github.com/ktorio/ktor/search?q=CurlIllegalStateException

Reply:
Prefer delicated exceptions. `CurlIllegalStateException` is deprecated (though there are no alternatives available)